src/main/scala/firrtl/passes/InferResets.scala

```diff
@@ -0,0 +1,200 @@
+        val target = toTarget(c.main, mod.name)(expr)
+        Utils.kind(expr) match {
+          case InstanceKind =>
+            val mod = instMap(target.ref)
+            val port = target.component.head match {
+              case TargetToken.Field(name) => name
+            }
+            target.copy(module = mod, ref = port, component = target.component.tail)
+          case _ => target
+        }
+      }
+      def onStmt(map: DriverMap)(stmt: Statement): Unit = {
+        stmt match {
+          case Connect(_, lhs, rhs) if lhs.tpe == ResetType =>
+            val target = makeTarget(lhs)
+            val driver = rhs.tpe match {
+              case ResetType => TargetDriver(makeTarget(rhs))
+              case tpe       => TypeDriver(tpe, () => makeTarget(rhs))
+            }
+            map(target) = driver :: Nil
+          case WDefInstance(_, inst, module, _) =>
+            instMap += (inst -> module)
+          case Conditionally(_, _, con, alt) =>
+            val conMap = new DriverMap
+            val altMap = new DriverMap
+            onStmt(conMap)(con)
+            onStmt(altMap)(alt)
+            // Default to outerscope if not found in alt
+            val altLookup = altMap.orElse(map).lift
+            for (key <- conMap.keys ++ altMap.keys) {
+              map(key) = (conMap.get(key).toList ++ altLookup(key).toList).flatten
+            }
+          case other => other.foreach(onStmt(map))
+        }
+      }
+      val types = new DriverMap
+      mod.foreach(onStmt(types))
+      types
+    }
+    c.modules.foldLeft(Map[ReferenceTarget, List[ResetDriver]]()) {
+      case (map, mod) => map ++ onMod(mod)
+    }
+  }
+
+  // Determine the type driving a given ResetType
+  private def resolve(map: Map[ReferenceTarget, List[ResetDriver]]): Try[Map[ReferenceTarget, Type]] = {
+    val res = mutable.Map[ReferenceTarget, Type]()
+    val errors = new Errors
+    def rec(target: ReferenceTarget): Type = {
+      val drivers = map(target)
+      res.getOrElseUpdate(target, {
+        val tpes = drivers.map {
+          case TargetDriver(t) => TypeDriver(rec(t), () => t)
+          case td: TypeDriver => td
+        }.groupBy(_.tpe)
+        if (tpes.keys.size != 1) {
+          // Multiple types of driver!
+          errors.append(DifferingDriverTypesException(target, tpes.toSeq))
+        }
+        tpes.keys.head
+      })
+    }
+    for ((target, _) <- map) {
+      rec(target)
+    }
+    Try {
+      errors.trigger()
+      res.toMap
+    }
+  }
+
+  private sealed trait TypeTree
+  private case class BundleTree(fields: Map[String, TypeTree]) extends TypeTree
+  private case class VectorTree(subType: TypeTree) extends TypeTree
+  // TODO ensure is only AsyncResetType or BoolType
+  private case class GroundTree(tpe: Type) extends TypeTree
+
+  private object TypeTree {
+    // TODO make return Try[TypeTree]
+    def fromTokens(tokens: (Seq[TargetToken], Type)*): TypeTree = tokens match {
```

Comment:
Clarify what the tokens first and second element represent in comment, as well as the return value.

Reply:
From where this is call, I see this as converting a bunch of ReferenceTargets (and their type) which all target the subelements of one aggregate typed component. This then converts all of them into what the new aggregate type should be. Why did you need to create a new class, and not just use the ir.Type?

Reply:
In addition, this seems to be a very useful function, irregardless of the async reset. We should consider putting it in a standard place, not hidden away here.

Reply:
I'd recommend adding more comments to all of these private functions.

Reply:
I've added some notes. The purpose of this type hierarchy is to represent a lookup structure for setting types of leaf-level nodes in nested Types. I created the new type structure to make the code more self-documenting in that the structure of the lookup is pretty explicit.

It could be expressed very similarly using `ir.Type` and then performing some sort of "union" of the two structures. The thing I like about the private type structure is it only contains the necessary information whereas having to create Fields and stuff would require me to fill in lots of values that I don't know, nor care about, and it could be a bit misleading.